src/cached_object_store.rs

```diff
@@ -0,0 +1,1067 @@
+
+    async fn read_head(&self) -> object_store::Result<Option<(ObjectMeta, Attributes)>>;
+}
+
+#[derive(Debug)]
+struct FsCacheStorage {
+    root_folder: std::path::PathBuf,
+}
+
+impl LocalCacheStorage for FsCacheStorage {
+    fn entry(
+        &self,
+        location: &object_store::path::Path,
+        part_size: usize,
+    ) -> Box<dyn LocalCacheEntry> {
+        Box::new(FsCacheEntry {
+            root_folder: self.root_folder.clone(),
+            location: location.clone(),
+            part_size,
+        })
+    }
+}
+
+impl Display for FsCacheStorage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "FsCacheStorage({})", self.root_folder.display())
+    }
+}
+
+#[derive(Debug)]
+struct FsCacheEntry {
+    root_folder: std::path::PathBuf,
+    location: Path,
+    part_size: usize,
+}
+
+impl FsCacheEntry {
+    async fn atomic_write(&self, path: &std::path::Path, buf: Bytes) -> object_store::Result<()> {
+        let tmp_path = path.with_extension(format!("_tmp{}", self.make_rand_suffix()));
+
+        // ensure the parent folder exists
+        if let Some(folder_path) = tmp_path.parent() {
+            fs::create_dir_all(folder_path).await.map_err(wrap_io_err)?;
+        }
+
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(&tmp_path)
+            .await
+            .map_err(wrap_io_err)?;
+        file.write_all(&buf).await.map_err(wrap_io_err)?;
+        fs::rename(tmp_path, path).await.map_err(wrap_io_err)
+    }
+
+    // every origin file will be split into multiple parts, and all the parts will be saved in the same
+    // folder. the part file name is expected to be in the format of `_part{part_size}-{part_number}`,
+    // examples: /tmp/mydata.csv/_part1mb-000000001
+    fn make_part_path(
+        root_folder: std::path::PathBuf,
+        location: &Path,
+        part_number: usize,
+        part_size: usize,
+    ) -> std::path::PathBuf {
+        // containing the part size in the file name, allows user change the part size on
+        // the fly, without the need to invalidate the cache.
+        let part_size_name = if part_size % (1024 * 1024) == 0 {
+            format!("{}mb", part_size / (1024 * 1024))
+        } else {
+            format!("{}kb", part_size / 1024)
+        };
+        let suffix = format!("_part{}-{:09}", part_size_name, part_number);
+        let mut path = root_folder.join(location.to_string());
+        path.push(suffix);
+        path
+    }
+
+    fn make_head_path(root_folder: std::path::PathBuf, location: &Path) -> std::path::PathBuf {
+        let suffix = "_head".to_string();
+        let mut path = root_folder.join(location.to_string());
+        path.push(suffix);
+        path
+    }
+
+    fn make_rand_suffix(&self) -> String {
+        let mut rng = rand::thread_rng();
+        (0..6).map(|_| rng.sample(Alphanumeric) as char).collect()
```

Comment:
I think this means there's a 1-in-million chance of collision on temp files. This seems to low too me. Maybe use ULID, or `tempfile()` (in the root_folder), or make the length longer (24 gives us [~ the same probability as a ULID](https://pkg.go.dev/github.com/oklog/ulid/v2#section-readme)).

Reply:
i was afraid of `tempfile()` might delete the file when the handle is out of scope. the file is actually not `temp` but a staging file, after renaming, it's considered to live a long time until get envicted.

i'll change it into 24.